### DISCUSSION TITLE
Irradiance sensors - occluders or nor?

### DISCUSSION THREAD

**Discussion by isolin** [mitsuba-renderer/mitsuba3#376] [Q&A] [answered]
I am experimenting with generated scenes where most of the object are simple primitives like cylinders or rects. I am interested to get irradiance readings for most of the scene object, so most of them contain irradiancemeters.

I can't get the measured light right and I suspect that when I call `render` with the desired sensor index, the other irradiancemeters (i.e. their parent primitives) are not considered occluders. I noticed this at regions where dense placement of objects (that are irradiancemeter at the same time) creates a lot of shadows but sensor readings seem to ignore them and measure strong energy instead.

**Are objects that I turn to irradiancemeters occluders or not?**

**Comment by njroussel** [comment, marked as answer]
Hi @isolin 
I believe they still are occluders. I setup a scene with 3 cubes which are aligned  to verify this:
```python
import mitsuba as mi
from mitsuba import ScalarTransform4f as T

mi.set_variant('cuda_ad_rgb')

scene_description = {
    'type': 'scene',
    'emitter': {
        'type': 'cube',
        'to_world': T.translate(mi.ScalarPoint3f(-5, 0, 0)),
        'emitter' : {
            'type': 'area',
            'radiance': {
                'type': 'rgb',
                'value' : [0.9, 0.9, 0.9]
            }
        }
    },
    'irradiancemeter': {
        'type': 'cube',
        'to_world': T.translate(mi.ScalarPoint3f(5, 0, 0)),
        'sensor' : {
            'type': 'irradiancemeter',
            'film': {
                'type': 'hdrfilm',
                'width': 1,
                'height': 1
            }
        }
    },
    #'irradiancemeter_occluder': {
    #    'type': 'cube',
    #    'sensor' : {
    #        'type': 'irradiancemeter',
    #    }
    #},
    'integrator': {
        'type': 'path'
    }
}

scene = mi.load_dict(scene_description)

out = mi.render(scene, spp=1024, sensor=scene.sensors()[0])
print(out)  # Has value 0 if irradiancemeter_occluder is in the scene, othwise it's non-zero
```

**Reply by isolin** [reply]
Hi @njroussel! Thank you, that is a reasonable test scene! I have no OptiX over here, so I had to change the mode. Both `mi.set_variant('scalar_rgb')` and `mi.set_variant('llvm_ad_rgb')` work as expected.

I forgot to mention that I am using a directional light to imitate sun. In that case the first sensor returns `0` either way. Is there an issue with the directional light? Perhaps [this](https://github.com/mitsuba-renderer/mitsuba3/issues/365#issuecomment-1303919233)?
```py
import mitsuba as mi
mi.set_variant('scalar_rgb')

from mitsuba import ScalarTransform4f as T

scene_description = {
    'type': 'scene',
    'emitter': {
        'type': 'directional',
        'direction': [-1, 0, 0],
        'irradiance' : {
            'type': 'rgb',
            'value': 1.0
        }
    },
    'irradiancemeter': {
        'type': 'cube',
        'to_world': T.translate(mi.ScalarPoint3f(5, 0, 0)),
        'sensor' : {
            'type': 'irradiancemeter',
            'film': {
                'type': 'hdrfilm',
                'width': 1,
                'height': 1
            }
        }
    },
    # 'irradiancemeter_occluder': {
    #    'type': 'cube',
    #    'sensor' : {
    #        'type': 'irradiancemeter',
    #    }
    # },
    'integrator': {
        'type': 'path'
    }
}

scene = mi.load_dict(scene_description)

out = mi.render(scene, spp=1024, sensor=scene.sensors()[0])
print(out.array)  # Always 0 which is wrong
```

**Reply by njroussel** [reply]
The issue you linked is indeed related. However it is not an issue with the `directional` `Emitter` but rather the choice of `Integrator` (`path` here).
The `directional` emitter is a delta emitter, in the sense that it only emits light in a single direction. This is problematic because if you think of how a pathtracer works it is practically impossible to reach an emitter with the same direction in which it emits - so there is 0 light contribution along any ray.
The `ptracer` integrator should do the job: it will start tracing rays from the emitters, hence completely getting rid of the issue I mentionned above.

**Reply by isolin** [reply]
Thank you, indeed it works fine now.
